### ISSUE TITLE
[LanguageServices] Language services need to support )__VSHPROPID2.VSHPROPID_IntellisenseUnknown  for progression to work on fresh VS instance

### ISSUE THREAD

**Issue by abpiskunov** [dotnet/roslyn-project-system#500]
So it turns out now that the regression is in core language services not supporting hier.GetProperty((uint)VSConstants.VSITEMID.Root, (int)__VSHPROPID2.VSHPROPID_IntellisenseUnknown , which results with whole progression not working (including classes and methods under *.cs files) when solution is opened in a fresh VS instance.

There two different logics in progression to get projects into cache:
-	On fresh VS instance they get projects form solution and check if they are supported or not  - it is not. Then it check if there is a supported “Intellisense type” – a guid that progression recognizes. It returned me a collection of 4 guids:

http://index/#Microsoft.VisualStudio.Progression.CodeSchema/Utils/ProjectHelper.cs,337
 
-                intellisenseTypes        Count = 4        System.Collections.Generic.List<System.Guid>
+                [0]        {c7fedb89-b36d-4a62-93f4-dc7a95999921}        System.Guid    <- CSharp http://source.roslyn.io/#Microsoft.VisualStudio.LanguageServices/Guids.cs,26 
+                [1]        {12c8a7d2-4681-11d2-b48a-0000f87572eb}        System.Guid  <- VB
+                [2]        {12c8a7d2-4681-11d2-b48a-0000f87572eb}        System.Guid
+                [3]        {c7fedb89-b36d-4a62-93f4-dc7a95999921}        System.Guid
	However when it tries to match those guids with project it query )__VSHPROPID2.VSHPROPID_IntellisenseUnknown property from IVsHierarchy, which fails for core project at the moment: http://index/#Microsoft.VisualStudio.Progression.CodeSchema/Utils/ProjectHelper.cs,845
-	When solution is closed and reopened in the same VS instance progression has different code path for adding projects to the cache: it just adds project that it receives form OnProjectAdded event and does not do any IsSupportedChecks – that’s why it does work in this case for Dependencies node, but it still does not fork for *.cs files(!).

So it looks like core language services need to implement __VSHPROPID2.VSHPROPID_IntellisenseUnknown  for this scenario to work, since progression logic seems to be too complex to touch in this case

**Comment by srivatsn** [comment]
What is the net effect of this bug? I see the dependencies node show up for my projects with the new LS.

**Comment by abpiskunov** [comment]
graphnodes for deeper hierarchy are not showing up and class/methods nodes are not showing up for *.cs files at all

**Comment by abpiskunov** [comment]
![image](https://cloud.githubusercontent.com/assets/1592091/18793800/5267a66a-8171-11e6-92c3-e18910588172.png)

**Comment by davkean** [comment]
Thanks - I'll look into this.

**Comment by davkean** [comment]
Okay, I looked into this and I do not believe that we should be returning an instance of ICSharpProjectRoot - as we don't implement this since the rewrite of the language service. Every usage I saw of this outside of the language service integration, was using it solely to sniff C#/VB projects. The three usages were:

1) CSLanSvcProvider in progression to light up for C#
2) ProjectLanguageHelper to sniff C#/VB
3) CSharpGraphcProvider to light up for C#. 

This explains the logic as to why the C#/VB nodes under .cs/vb that aren't showing. It doesn't explain to me why Dependencies nodes aren't showing up. From what I can see, They just depend on the DependenciesTree capability. @abpiskunov What's cause this to know show up?

**Comment by abpiskunov** [comment]
@davkean i tried to explain the flow in my original repro , but was not clear I guess .. Let me try again:

- when new VS instance starts and progression initializes cache, it tries to get projects from solution and does check if project supported or not. 
- currently cs/vb core projects are unsupported because of that issue.
- thus when capability check happens on fresh vs instance - projects cache in progression is empty and no check happens at all

when you close and reopen solution in the same VS instance, cache is updated , since on reload their cache adds projects differently comparing to how they are added on initial cache creation time: they just listen OnProjectAdded event and add project to the cache without any compatibility checks. This time when capability check happens , our project is in cache and is checked for capability.

**Comment by davkean** [comment]
Can you point me the compatibility check? Given you've got a project capability - that should be the check, not whether it has a recognized VSHPROPID_IntellisenseUnknown or not.

**Comment by abpiskunov** [comment]
that is the problem - we don't have a project to check capability since it is not in the progression cache.

**Comment by davkean** [comment]
I'm asking, where this code lives:

> when new VS instance starts and progression initializes cache, it tries to get projects from solution and does check if project supported or no

**Comment by abpiskunov** [comment]
http://index/Microsoft.VisualStudio.GraphProviderPackage/R/528a66667f3e8fd8.html  its in this class , however index is not updated , check the actual class in vso

**Comment by davkean** [comment]
@abpiskunov and I talked offline, there's two issues:

1) The C#/VB graph providers need to be updated to use capabilities: https://github.com/dotnet/roslyn-project-system/issues/592.

2) The caching logic for projects should be updated to look for project's with matching capabilities.

**Comment by abpiskunov** [comment]
Fix checked in today
